File: devtrouble-ai-frontend/src/types/ai.ts
export interface ConversationMessage {
  role: 'user' | 'assistant';
  content: string;
}

export interface AiSearchRequest {
  query: string;
  project_id?: string;
  /** 이전 턴들. 서버는 대화를 저장하지 않으므로 매번 지금까지의 대화를 그대로 실어 보낸다. */
  history?: ConversationMessage[];
}

export interface AiCitation {
  document_id: string;
  title: string;
  relevance_score: number;
}

export interface AiSearchResponse {
  answer: string;
  cause: string | null;
  similar_cases: string | null;
  solution: string | null;
  citations: AiCitation[];
  /** 트러블슈팅과 무관한 질문(인사/잡담)으로 분류되었는지. false면 검색 없이 안내만 반환된 것. */
  on_topic: boolean;
  /** 답변이 검색된 문서에 실제로 근거했다고 자체 검증됐는지. false면 확신도가 낮을 수 있음. */
  is_grounded: boolean;
}

/** /ai/search/stream(SSE)이 보내는 이벤트. stage="token"일 때 answer(부분 답변), "done"일 때 response가 채워진다. */
export interface AiSearchStreamEvent {
  stage: string;
  answer?: string;
  response?: AiSearchResponse;
}

File: devtrouble-ai-backend/app/schemas/ai.py
from typing import Literal

from pydantic import BaseModel, Field


class ConversationMessage(BaseModel):
    """멀티턴 대화 이력의 한 턴. 클라이언트가 이전 턴들을 그대로 담아 보낸다 (서버 세션 저장 없음)."""

    role: Literal["user", "assistant"]
    content: str


class AiSearchRequest(BaseModel):
    query: str = Field(min_length=1, max_length=1000)
    project_id: str | None = None  # 특정 프로젝트로 검색 범위를 좁히고 싶을 때
    # 후속 질문("그럼 그 경우엔?")을 이해하는 데 쓰인다. 서버는 대화를 저장하지 않으므로
    # 매 요청마다 클라이언트가 지금까지의 턴을 그대로 실어 보내야 한다.
    history: list[ConversationMessage] = Field(default_factory=list)


class AiCitation(BaseModel):
    document_id: str
    title: str
    relevance_score: float


class AiSearchResponse(BaseModel):
    """FR-AI-04/05: 원인 / 유사 사례 / 해결 방법 + 출처."""

    answer: str
    cause: str | None = None
    similar_cases: str | None = None
    solution: str | None = None
    citations: list[AiCitation] = Field(default_factory=list)
    # 이 질문이 트러블슈팅 관련으로 분류되었는지 (classify 노드 결과).
    on_topic: bool = True
    # 자체 검증(validate 노드)에서 "Context에 근거했다"고 판단됐는지.
    # False라면 재시도 예산을 다 썼는데도 여전히 근거가 불확실하다는 뜻이니
    # 프론트에서 "확실하지 않을 수 있습니다" 같은 경고를 붙이는 데 쓸 수 있다.
    is_grounded: bool = True
